### DISCUSSION TITLE
Kasa Devices

### DISCUSSION THREAD

**Discussion by grahnema** [theyosh/TerrariumPI#625] [General]
Just a word of warning to anyone using Kasa Smart strips to avoid firmware upgrades and possibly Kasa all together for Terrarium Pi.  For the life of me I couldn't get my RPi3B running TPi4 to control the Kasa HS300 smart strip other than plug 1. After looking into it for quite some time it seems there was a recent firmware update that kills local only control forcing all commands through the cloud.   There was one article that said you could skip the create account step to bypass the cloud and add the strip to your local wifi only but that no longer works either, the set up process just loops back forcing you to the account creation step.     I tried a lot of plug number options after the IP: 
xx.xx.xx.xx,1
xx.xx.xx.xx:1
xx.xx.xx.xx:01
xx.xx.xx.xx,01
xx.xx.xx.xx(1)
xx.xx.xx.xx(01)
xx.xx.xx.xx,(1)
xx.xx.xx.xx,(01) 
I even tried adding the port (:9999) after the IP followed by all the plug number combinations as well but the only thing that works intermittently I should add, is just the IP of the strip which again only controls the first plug.

**Comment by biohazard827** [comment]
I've never gotten my HS300 to work. When I manually add the address with [IP Address],[plug number] I get the following:

```
2021-12-02 16:01:37,130 - WARNING - hardware.relay        - Error getting new data from relay Kasa Smart relay named 'UV Lamp' at address '10.0.0.174,1'. Error: unknown, retrying in 0.5 seconds...
AttributeError: 'SmartStrip' object has no attribute 'plugs'
    plugs = [self.device] if len(self._address) == 1 else self.device.plugs
  File "/home/pi/TerrariumPI/hardware/relay/kasa_relay.py", line 64, in __get_hardware_state
    raise self._exception
  File "/usr/lib/python3.7/concurrent/futures/_base.py", line 384, in __get_result
    return self.__get_result()
  File "/usr/lib/python3.7/concurrent/futures/_base.py", line 432, in result
    data = toggle.result()
  File "/home/pi/TerrariumPI/hardware/relay/kasa_relay.py", line 75, in _get_hardware_value
Traceback (most recent call last):
2021-12-02 16:01:37,107 - ERROR   - hardware.relay.kasa_relay - 'SmartStrip' object has no attribute 'plugs'
```

I checked that python-kasa is able to see the device by downloading the python-kasa source and running 'kasa discover' and I get these results:

```
pi@raspberrypi:~/TerrariumPI $ kasa --host 10.0.0.174
No --strip nor --bulb nor --plug given, discovering..
== TP-LINK_Power Strip_6057 - HS300(US) ==
        Host: 10.0.0.174
        Device state: OFF

        == Plugs ==
        * Socket 'UV Lamp' state: OFF on_since: None
        * Socket 'Heat Lamp' state: OFF on_since: None
        * Socket 'Fan' state: OFF on_since: None
        * Socket 'Plug 4' state: OFF on_since: None
        * Socket 'Plug 5' state: OFF on_since: None
        * Socket 'Dehumidifier' state: OFF on_since: None

        == Generic information ==
        Time:         2021-12-02 16:05:21
        Hardware:     2.0
        Software:     1.0.3 Build 201203 Rel.165457
        MAC (rssi):   C0:06:C3:42:60:57 (-42)
        Location:     {'latitude': 29.7846, 'longitude': -95.7733}

        == Device specific information ==
        LED state: True
        Childs count: 6
        On since: None


        == Current State ==
        <EmeterStatus power=0.0 voltage=120.025 current=0.0 total=0.0>
```
So I would think the web server would be able to see it.

**Reply by theyosh** [reply]
Hmm, yeah, this is never tested in TP4. Maybe that was not clear. And I do not own this kind of hardware. So the rewrite was a bit guessing.

But if you have some time to do some debugging, maybe we can fix this. I will look in the discover part of the code, to see if I can match that also... But I have to do it without the actual hardware.. will be a bit harder to do.

Or, if you are handy enough, and willing, setup a port forward to your PI and create an ssh account for me. That way I can create and test the code on your Pi directly. That is always the fastest way to do. But, you need to trust me, as you create a login on your Pi on your home network. Just to be clear.

If you think you can do this, please send login detail to: github@theyosh.nl so that they do not end up here.

**Reply by grahnema** [reply]
I sent my HS300 back because no matter what I tried I could only get one outlet to work and I believe it has something to do with TP link killing local control as part of a security measure.  The single outlet plugs and even switches still seem to work fine but multiple outlet devices such as the HS300 and KP303 are having issues responding to commands to control more than one outlet. It seems to be that as commands would have normally gone directly to the device locally over your home network the socket number is being filtered or blocked out when it goes through the TP-Link cloud. 

However I did find one lead that may be promising when it comes to controlling multiple outlet devices over local Wi-Fi but its a complete app that can be found here https://github.com/DaveGut/HubitatActive.  It may be as simple as using that app rather than TP links to configure the strip to work locally for it to work with TP4.

**Reply by theyosh** [reply]
Hmm, are you sure your device was/is broken? As it turned out that my code was not correct also. So as far I can judge, my code should work now. But I can't tell about the TP Kasa firmware updates. I have no information about that...

**Comment by biohazard827** [comment]
I did some debugging and it looks like python-kasa changed the power strip object to have 'children' instead of 'plugs'. I made the changes to the file but do not have permissions to push them (rightfully so). I have tested locally and I am now able to scan and have all the relays added automatically as expected. Adding them manually also works. This was tested with the HS300 as well as my outdoor two outlet KP400. I have attached the kasa_relay file as a TXT here.

[kasa_relay.txt](https://github.com/theyosh/TerrariumPI/files/7650604/kasa_relay.txt)

**Reply by theyosh** [reply]
Whooott!! This is nice! I will update my code with this. Thanks for testing and debugging. Makes my life a bit easier :P

**Reply by biohazard827** [reply]
NP, thanks for the software in general!

**Reply by theyosh** [reply]
So I updated the code. Could you see if that is still working for you? Then I know that it is all fine now.

**Reply by biohazard827** [reply]
I pulled the latest and everything is still working as expected.
